### PR TITLE
WIP: Add gamma family functions from SciPy with C code for GPU

### PR DESCRIPTION
**Pull request by domenzain**
The C code is MIT-licensed.

### CODE REVIEW

**Comment by domenzain** [comment]
Hi @nouiz @twiecki,

Here is a first draft for the scalar functions.
This takes the work in Theano/Theano_lgpl, but with the MIT-relicensed C code and @nouiz's commits replayed on top.

With a couple of wrappers I wrote for the other gamma functions.
The idea is to use this in PyMC3 for CDF functions for the Gamma family of distributions.

Do I need to do something special for the scalars to become Elementwise tensor operations?

Thanks,

**Comment by abergeron** [comment]
To make this usable on tensor you have to define the wrappers in tensor/basic.py (search for _scal_elemwise).

Otherwise, you have a flake8 error in scalar/basic.py

File: theano/scalar/c_code/gamma.c [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+    + 76.18009172947146      /(n+1)
+    - 86.50532032941677      /(n+2)
+    + 24.01409824083091      /(n+3)
+    -  1.231739572450155     /(n+4)
+    +  0.1208650972866179e-2 /(n+5)
+    -  0.5395239384953e-5    /(n+6);
+  return (n+0.5) *log((n+5.5)/LN_BASE) +(LN_SQRT_2PI +log(s/n) -5.0);
+}  /* logGamma() */
+
+#else /*--------------------------------------------------------------*/
```
[the remark this answers is not in the document]
Comment (domenzain) [reply theano/scalar/c_code/gamma.c]: Will do.

File: theano/scalar/c_code/gamma.c [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+    if (fabs(d) < fabs(a)) dx += d;
+    if (x +dx > 0) x += dx;
+    else           x /= 2;
+  } while (fabs(a) > 1e-10 *x);
+  if (fabs(dp) > EPS_QTL *prob) return -1;
+  return x *theta;              /* check for convergence and */
+}  /* GammaqtlQ() */            /* return the computed quantile */
+
+#endif
+/*--------------------------------------------------------------------*/
```
[the remark this answers is not in the document]
Comment (domenzain) [reply theano/scalar/c_code/gamma.c]: Will do.

**Comment by domenzain** [comment]
Hm, I don't think the scalar/basic.py flake8 errors come from any of the modifications in this PR.
I think it would be more sensible to correct those separately.

File: theano/scalar/c_code/gamma.c [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+    -  1.231739572450155     /(n+4)
+    +  0.1208650972866179e-2 /(n+5)
+    -  0.5395239384953e-5    /(n+6);
+  return (n+0.5) *log((n+5.5)/LN_BASE) +(LN_SQRT_2PI +log(s/n) -5.0);
+}  /* logGamma() */
+
+#else /*--------------------------------------------------------------*/
+
+DEVICE double logGamma (double n)
+{                               /* --- compute ln(Gamma(n))         */
```
[the remark this answers is not in the document]
Comment (domenzain) [reply theano/scalar/c_code/gamma.c]: These behave exactly like those in Theano/Theano_lgpl.
I imagine is still is better than nothing, depending on the size of the calculation.

Could you point me to an example of libgpuarray dispatch?

**Comment by abergeron** [comment]
I don't have any objections to this code going in as-is, but I will leave someone else to do the merge.

@twiecki

**Comment by domenzain** [comment]
@twiecki, I think this is useable as is given the tensor extension.
And indeed will mostly be used for the CPU.

I'll have a look at what it takes to make libgpuarray map a function to the different back-ends.

**Comment by twiecki** [comment]
@domenzain Should we merge this now or do you want to think about adding it here?

**Comment by domenzain** [comment]
I think this is ready to merge.
If the libgpuarray ops become available I would make a new PR to update these scalar operations with them.

**Comment by twiecki** [comment]
Thanks @domenzain!

**Pull request merged**